framework/src/main/java/org/checkerframework/framework/util/SimpleQualifierHierarchy.java

```diff
@@ -0,0 +1,287 @@
+
+    /** {@link QualifierKindHierarchy}. */
+    protected final QualifierKindHierarchy qualifierKindHierarchy;
+
+    /** Set of top annotation mirrors. */
+    protected final Set<AnnotationMirror> tops;
+
+    /** Set of bottom annotation mirrors. */
+    protected final Set<AnnotationMirror> bottoms;
+
+    /** Mapping from {@link QualifierKind} to its corresponding {@link AnnotationMirror}. */
+    protected final Map<QualifierKind, AnnotationMirror> qualifierMap;
+
+    /** Set of all annotations in the hierarchy. */
+    protected final Set<? extends AnnotationMirror> qualifiers;
+
+    /**
+     * Creates a type hierarchy from the given classes.
+     *
+     * @param qualifierClasses classes of annotations that are the qualifiers for this hierarchy
+     * @param elements element utils
+     */
+    public SimpleQualifierHierarchy(
+            Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+        this.qualifierKindHierarchy = createQualifierKindHierarchy(qualifierClasses);
+
+        this.qualifierMap = createAnnotationMirrors(elements);
+        Set<AnnotationMirror> qualifiers = AnnotationUtils.createAnnotationSet();
+        qualifiers.addAll(qualifierMap.values());
+        this.qualifiers = Collections.unmodifiableSet(qualifiers);
+
+        this.tops = createTops();
+        this.bottoms = createBottoms();
+    }
+
+    /**
+     * Create the {@link QualifierKindHierarchy}. (Subclasses may override to return a subclass of
+     * QualifierKindHierarchy.)
+     *
+     * @param qualifierClasses classes of annotations that are the qualifiers for this hierarchy
+     * @return the newly created qualifier kind hierarchy
+     */
+    protected QualifierKindHierarchy createQualifierKindHierarchy(
+            Collection<Class<? extends Annotation>> qualifierClasses) {
+        return new QualifierKindHierarchy(qualifierClasses);
+    }
+
+    /**
+     * Creates and returns a mapping from qualifier kind to an annotation mirror created from the
+     * qualifier kind's annotation class.
+     *
+     * @param elements element utils
+     * @return a mapping from qualifier kind to an annotation mirror created from the qualifier
+     *     kind's annotation class
+     */
+    protected Map<QualifierKind, AnnotationMirror> createAnnotationMirrors(Elements elements) {
+        Map<QualifierKind, AnnotationMirror> quals = new TreeMap<>();
+        for (QualifierKind kind : qualifierKindHierarchy.allQualifierKinds()) {
+            if (kind.hasElements()) {
+                throw new UserError(
+                        "SimpleQualifierHierarchy cannot be used with annotations that have elements. Found %s: ",
+                        kind);
+            }
+            quals.put(kind, AnnotationBuilder.fromClass(elements, kind.getAnnotationClass()));
+        }
+        return Collections.unmodifiableMap(quals);
+    }
+
+    protected Set<AnnotationMirror> createTops() {
```

Comment:
Please write Javadoc documentation for all methods.
The next two methods need it, too.